### PR TITLE
Correctly validate ACME PoP against public key

### PR DESCRIPTION
**Pull request by cipherboy**
ACME's proof of possession based revocation uses a signature from the private key, but only sends the public copy along with the request. Ensure the public copy matches the certificate, instead of failing to cast it to a private key.

### CODE REVIEW

**Review by cipherboy — COMMENTED** [review]
Tests look good, thanks @stevendpclark!

**Review by stevendpclark — APPROVED** [review]
This review is just for the non-test code.

File: builtin/logical/pki/path_acme_revoke.go [hunk 1]
```diff
@@ -136,14 +136,14 @@ func (b *backend) acmeRevocationHandler(acmeCtx *acmeContext, _ *logical.Request
 func (b *backend) acmeRevocationByPoP(acmeCtx *acmeContext, userCtx *jwsCtx, cert *x509.Certificate, config *crlConfig) (*logical.Response, error) {
 	// Since this account does not exist, ensure we've gotten a private key
 	// matching the certificate's public key.
-	signer, ok := userCtx.Key.Key.(crypto.Signer)
+	public, ok := userCtx.Key.Key.(crypto.PublicKey)
```
Comment (kitography) [inline builtin/logical/pki/path_acme_revoke.go:139]: This comment is now very confusing.
I think what needs to happen is to check: "VerifyJWS(signature string)"  - that does seem to happen from the following call structure:
- acmeRevocationByPoP is called by: acmeRevocationHandler
- acmeRevocationHandler is called by patternAcmeRevoke
- patternAcmeRevoke wraps acmeRevocationHandler in acmeParsedWrappers
- acmeParsedWrappers calls ParseRequestParams
- ParseRequestParams calls VerifyJWS
- If VerifyJWS returns an error, then the call should be returned there.

This is confusing enough that I'd totally make a mistake with it.

File: builtin/logical/pki/path_acme_revoke.go [hunk 2]
```diff
@@ -136,14 +136,14 @@ func (b *backend) acmeRevocationHandler(acmeCtx *acmeContext, _ *logical.Request
...
 	if !ok {
 		return nil, fmt.Errorf("unable to revoke certificate: unable to parse JWS key of type (%T): %w", userCtx.Key.Key, ErrMalformed)
 	}
 
 	// Ensure that our PoP is indeed valid.
```
Comment (kitography) [inline builtin/logical/pki/path_acme_revoke.go:144]: comment above could say something about we have proven possession of the key in the certificate; "public" is a really hard variable to use here - and more specific (proofKey or certKey?) would be nicer - and which is which on the error message looks confusing.

File: builtin/logical/pki/path_acme_test.go [hunk 3]
```diff
@@ -1405,6 +1428,131 @@ func TestAcmeValidationError(t *testing.T) {
...
+	certResp, err = vaultClient.Logical().ReadWithContext(ctx, "pki/cert/"+serialFromCert(acmeCert))
+	require.NoError(t, err, "failed to read certificate status")
+	require.NotNil(t, certResp, "certificate status response was nil")
+	revocationTime = certResp.Data["revocation_time"].(json.Number)
+	revocationTimeInt, err = revocationTime.Int64()
+	require.NoError(t, err, "failed converting revocation_time value: %v", revocationTime)
+	require.Greater(t, revocationTimeInt, int64(0),
+		"revocation time was not greater than 0, cert was not revoked: %v", revocationTimeInt)
+
+	// Make sure we can revoke a certificate without a registered ACME account
```
Comment (kitography) [inline builtin/logical/pki/path_acme_test.go:1496]: isn't acmeClient1 a registered ACME account?  Or does that deregister when we revoke it's cert, above.

File: builtin/logical/pki/path_revoke.go [hunk 4]
```diff
@@ -461,26 +479,23 @@ func validatePrivateKeyMatchesCert(signer crypto.Signer, certReference *x509.Cer
 	// See: https://github.com/golang/go/blob/c6a2dada0df8c2d75cf3ae599d7caed77d416fa2/src/crypto/tls/tls.go#L304-L331
 	switch certPub := certReference.PublicKey.(type) {
 	case *rsa.PublicKey:
-		privPub, ok := signer.Public().(*rsa.PublicKey)
+		privPub, ok := verifier.(*rsa.PublicKey)
 		if !ok {
 			return errutil.UserError{Err: "provided private key type does not match certificate's public key type"}
```
Comment (kitography) [inline builtin/logical/pki/path_revoke.go:484]: this is now under validatePublicKeyMatchesCert - the error message refers to a private key.

File: builtin/logical/pki/path_revoke.go [hunk 5]
```diff
@@ -461,26 +479,23 @@ func validatePrivateKeyMatchesCert(signer crypto.Signer, certReference *x509.Cer
...
 		}
-		if err := signer.(*rsa.PrivateKey).Validate(); err != nil {
-			return err
-		}
 		if certPub.N.Cmp(privPub.N) != 0 || certPub.E != privPub.E {
 			return errutil.UserError{Err: "provided private key does not match certificate's public key"}
```
Comment (kitography) [inline builtin/logical/pki/path_revoke.go:487]: and again below here - privateKey shouldn't be referenced in publicKey Function?

**Review by kitography — COMMENTED** [review]
I don't think that this is wrong, but it is really, really confusing to read.

File: builtin/logical/pki/path_revoke.go (hunk 4 above)
Comment (cipherboy) [reply builtin/logical/pki/path_revoke.go:484]: `validatePrivateKeyMatchesCert` calls `validatePublicKeyMatchesCert` and most callers outside of ACME are using an explicit private key. Here, in ACME, we're validating the private key by proxy (JWS signature over the message).

Comment (cipherboy) [reply builtin/logical/pki/path_revoke.go:484]: I've updated the `acmeRevocationByPoP` error message to say `unable to verify PoP of private key provided by proxy: ` to clarify to the caller why private key is mentioned elsewhere in the error message.

**Review by kitography — APPROVED** [review]
LGTM

**Pull request merged**